packages/block-library/src/query-loop/edit.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useMemo } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import {
+	BlockContextProvider,
+	InnerBlocks,
+	BlockPreview,
+} from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { useQueryContext } from '../query';
+
+const TEMPLATE = [ [ 'core/post-title' ], [ 'core/post-content' ] ];
+export default function QueryLoopEdit( { clientId, context: { query } } ) {
+	const [ { page } ] = useQueryContext();
+	const [ activeBlockContext, setActiveBlockContext ] = useState();
+
+	const { posts, blocks } = useSelect(
+		( select ) => ( {
+			posts: select( 'core' ).getEntityRecords( 'postType', 'post', {
+				...query,
+				offset: query.per_page * ( page - 1 ) + query.offset,
+				page,
+			} ),
+			blocks: select( 'core/block-editor' ).getBlocks( clientId ),
+		} ),
+		[ query, page, clientId ]
+	);
+
+	const blockContexts = useMemo(
+		() =>
+			posts?.map( ( post ) => ( {
+				postType: post.type,
+				postId: post.id,
+			} ) ),
+		[ posts ]
+	);
+	return blockContexts
+		? blockContexts.map( ( blockContext ) => (
+				<BlockContextProvider
+					key={ blockContext.postId }
+					value={ blockContext }
+				>
+					{ blockContext ===
+					( activeBlockContext || blockContexts[ 0 ] ) ? (
```

Comment:
Could alternatively use the block context index for comparison, which might be a bit safer (since we're comparing objects here with `===`; which is probably fine due to the way they're memo-ized and set via `setActiveBlockContext`), but it might be overall a bit less fragile. (Could probably even use the post ID instead? Those are unique even across post types IIRC.)

That would also allow to turn `activeBlockContext` state into a number, initialize with `0`, and handle the `|| blockContexts[ 0 ]` fallback a bit more elegantly.

Reply:
I don't see why this is more fragile.

I do see it will be simpler because of the initialization to 0, but ideally, there shouldn't be an initial active block context, and we should unset them on blur. I was just running into a bug with `InnerBlocks` that should be fixed in its soon to land refactor.

Reply:
> I don't see why this is more fragile.

Future contributors to this code might easily miss that we're comparing objects by-reference here. If they make a change that causes that referential equality to be no longer valid, this comparison will break.